src/new-storage/DeleteExpiredCmd.js

```diff
@@ -0,0 +1,139 @@
+
+        const authProvider = new cassandra.auth.PlainTextAuthProvider(config.cassandraNew.username, config.cassandraNew.password)
+        this.cassandraClient = new cassandra.Client({
+            contactPoints: [...config.cassandraNew.hosts],
+            localDataCenter: config.cassandraNew.datacenter,
+            keyspace: config.cassandraNew.keyspace,
+            authProvider,
+            pooling: {
+                maxRequestsPerConnection: 32768
+            }
+        })
+
+        // used for limited concurrency
+        this.limit = pLimit(5)
+    }
+
+    async _getStreams() {
+        const result = []
+
+        const query = 'SELECT DISTINCT stream_id, partition FROM bucket'
+        const resultSet = await this.cassandraClient.execute(query)
+
+        if (resultSet.rows.length) {
+            resultSet.rows.forEach((row) => {
+                result.push({
+                    streamId: row.stream_id,
+                    partition: row.partition
+                })
+            })
+        }
+
+        return result
+    }
+
+    async _fetchStreamsInfo(streams) {
+        const tasks = streams.map((stream) => {
+            return this.limit(async () => {
+                const url = `${this.baseUrl}/api/v1/streams/${stream.streamId}/validation`
+                return fetch(url).then((res) => res.json()).then((json) => {
+                    return {
+                        streamId: stream.streamId,
+                        partition: stream.partition,
+                        storageDays: parseInt(json.storageDays)
+                    }
+                }).catch((e) => console.error(e))
+            })
+        })
+
+        return Promise.all(tasks)
+    }
+
+    async _deleteExpired(streamsInfo) {
+        const tasks = streamsInfo.map((stream) => {
+            const { bucketId, dateCreate, streamId, partition } = stream
+            const queries = [
+                {
+                    query: 'DELETE FROM bucket WHERE stream_id = ? AND partition = ? AND date_create = ?',
+                    params: [streamId, partition, dateCreate]
+                },
+                {
+                    query: 'DELETE FROM stream_data WHERE stream_id = ? AND partition = ? AND bucket_id = ?',
+                    params: [streamId, partition, bucketId]
+                }
+            ]
+
+            return this.limit(async () => {
+                await this.cassandraClient.batch(queries, {
+                    prepare: true
+                }).catch((err) => {
+                    console.error(err)
+                })
+            })
+        })
+
+        return Promise.all(tasks)
+    }
+
+    async _getExpiredBuckets(streamsInfo) {
+        const result = []
+
+        const query = 'SELECT * FROM bucket WHERE stream_id = ? AND partition = ? AND date_create <= ?'
+
+        const tasks = streamsInfo.map((stream) => {
+            const { streamId, partition, storageDays } = stream
```

Comment:
if `_fetchStreamsInfo` errors then `stream` will be `undefined` and this will give a destructuring error

Reply:
To protect against this I suggest adding `streamsInfo.filter(Boolean).map(…)` here and the `_deleteExpired` calls, or doing the same to the resolved values for `_fetchStreamsInfo` and here in `_getExpiredBuckets`.